modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -952,4 +983,30 @@ public interface ImplementorFactory<RowT> {
         /** Creates the relational node implementor with the given context. */
         LogicalRelImplementor<RowT> create(ExecutionContext<RowT> ctx, ResolvedDependencies resolvedDependencies);
     }
+
+    static final class FetchColocationGroups {
+
+        private final ResolvedDependencies deps;
+
+        FetchColocationGroups(ResolvedDependencies deps) {
+            this.deps = deps;
+        }
+
+        CompletableFuture<Map<Integer, ColocationGroup>> execute() {
+            List<CompletableFuture<Pair<Integer, ColocationGroup>>> list = new ArrayList<>();
+
+            for (Integer tableId : deps.tableIds()) {
+                CompletableFuture<ColocationGroup> f = deps.fetchColocationGroup(tableId);
+                list.add(f.thenApply(c -> new Pair<>(tableId, c)));
+            }
+
+            CompletableFuture<Void> all = CompletableFuture.allOf(list.toArray(new CompletableFuture[0]));
+
+            return all.thenApply(
+                    v -> list.stream()
+                            .map(CompletableFuture::join)
+                            .collect(Collectors.toMap(Pair::getFirst, Pair::getSecond)));
+
+        }
+    }
```

Comment:
Seems, this method can be static helper method and class can be dropped.
Or maybe the whole method can be moved to `ResolvedDependencies` class.

Reply:
Sure. This class can be replaced with a simple function that does exactly what execute does.

Reply:
Fixed.